File: src/main/java/com/tapadoo/slacknotifier/SlackServerAdapter.java
package com.tapadoo.slacknotifier;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import jetbrains.buildServer.issueTracker.Issue;
import jetbrains.buildServer.serverSide.*;
import jetbrains.buildServer.serverSide.settings.ProjectSettingsManager;
import jetbrains.buildServer.users.SUser;
import jetbrains.buildServer.users.UserSet;
import jetbrains.buildServer.vcs.SVcsModification;
import jetbrains.buildServer.vcs.SelectPrevBuildPolicy;
import org.joda.time.Duration;
import org.joda.time.format.PeriodFormatter;
import org.joda.time.format.PeriodFormatterBuilder;

import java.io.BufferedOutputStream;
import java.io.IOException;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.Collection;
import java.util.Date;
import java.util.List;


/**
 * Created by jasonconnery on 02/03/2014.
 */
public class SlackServerAdapter extends BuildServerAdapter {

    private final SBuildServer buildServer;
    private final SlackConfigProcessor slackConfig;
    private final ProjectSettingsManager projectSettingsManager;
    private final ProjectManager projectManager ;
    private Gson gson ;

    public SlackServerAdapter(SBuildServer sBuildServer, ProjectManager projectManager, ProjectSettingsManager projectSettingsManager , SlackConfigProcessor configProcessor) {

        this.projectManager = projectManager ;
        this.projectSettingsManager = projectSettingsManager ;
        this.buildServer = sBuildServer ;
        this.slackConfig = configProcessor ;

    }

    public void init()
    {
        buildServer.addListener(this);
    }

    private Gson getGson()
    {
        if( gson == null )
        {
            gson = new GsonBuilder().create() ;
        }

        return gson ;
    }

    @Override
    public void buildStarted(SRunningBuild build) {
        super.buildStarted(build);

        if( !build.isPersonal() && slackConfig.postStarted() )
        {
            postStartedBuild(build);
        }
    }

    @Override
    public void buildFinished(SRunningBuild build) {
        super.buildFinished(build);

        if( !build.isPersonal() && build.getBuildStatus().isSuccessful() && slackConfig.postSuccessful() )
        {
            processSuccessfulBuild(build);
        }
        else if ( !build.isPersonal() && build.getBuildStatus().isFailed() && slackConfig.postFailed() )
        {
            postFailureBuild(build);
        }
        else
        {
            //TODO - modify in future if we care about other states
        }
    }

    private void postStartedBuild(SRunningBuild build )
    {
        //Could put other into here. Agents maybe?
        String message = String.format("Project '%s' build started." , build.getFullName());
        postToSlack(build, message, true);
    }
    private void postFailureBuild(SRunningBuild build )
    {
        String message = "";

        PeriodFormatter durationFormatter = new PeriodFormatterBuilder()
                .printZeroRarelyFirst()
                .appendHours()
                .appendSuffix(" hour", " hours")
                .appendSeparator(" ")
                .printZeroRarelyLast()
                .appendMinutes()
                .appendSuffix(" minute", " minutes")
                .appendSeparator(" and ")
                .appendSeconds()
                .appendSuffix(" second", " seconds")
                .toFormatter();

        Duration buildDuration = new Duration(1000*build.getDuration());

        message = String.format("Project '%s' build failed! ( %s )" , build.getFullName() , durationFormatter.print(buildDuration.toPeriod()));

        postToSlack(build, message, false);
    }

    private void processSuccessfulBuild(SRunningBuild build) {

        String message = "";

        PeriodFormatter durationFormatter = new PeriodFormatterBuilder()
                    .printZeroRarelyFirst()
                    .appendHours()
                    .appendSuffix(" hour", " hours")
                    .appendSeparator(" ")
                    .printZeroRarelyLast()
                    .appendMinutes()
                    .appendSuffix(" minute", " minutes")
                    .appendSeparator(" and ")
                    .appendSeconds()
                    .appendSuffix(" second", " seconds")
                    .toFormatter();

        Duration buildDuration = new Duration(1000*build.getDuration());

        message = String.format("Project '%s' built successfully in %s." , build.getFullName() , durationFormatter.print(buildDuration.toPeriod()));

        postToSlack(build, message, true);
    }

    /**
     * Post a payload to slack with a message and good/bad color. Committer summary is automatically added as an attachment
     * @param build the build the message is relating to
     * @param message main message to include, 'Build X completed...' etc
     * @param goodColor true for 'good' builds, false for danger.
     */
    private void postToSlack(SRunningBuild build, String message, boolean goodColor) {
        try{

            URL url = new URL(slackConfig.getPostUrl());

            SlackProjectSettings projectSettings = (SlackProjectSettings) projectSettingsManager.getSettings(build.getProjectId(),"slackSettings");

            if( ! projectSettings.isEnabled() )
            {
                return ;
            }

            String iconUrl = projectSettings.getLogoUrl();

            if(iconUrl == null || iconUrl.length() < 1 )
            {
                iconUrl = slackConfig.getLogoUrl() ;
            }

            String configuredChannel = build.getParametersProvider().get("SLACK_CHANNEL");
            String channel = this.slackConfig.getDefaultChannel();

            if( configuredChannel != null && configuredChannel.length() > 0 )
            {
                channel = configuredChannel ;
            }
            else if( projectSettings != null && projectSettings.getChannel() != null && projectSettings.getChannel().length() > 0 )
            {
                channel = projectSettings.getChannel() ;
            }

            UserSet<SUser> committers = build.getCommitters(SelectPrevBuildPolicy.SINCE_LAST_BUILD);
            StringBuilder committersString = new StringBuilder();

            for( SUser committer : committers.getUsers() )
            {
                if( committer != null)
                {
                    String committerName = committer.getName() ;
                    if( committerName == null || committerName.equals("") )
                    {
                        committerName = committer.getUsername() ;
                    }

                    if( committerName != null && !committerName.equals(""))
                    {
                        committersString.append(committerName);
                        committersString.append(",");
                    }
                }
            }

            if( committersString.length() > 0 )
            {
                committersString.deleteCharAt(committersString.length()-1); //remove the last ,
            }

            String committersMsg = committersString.toString();


            JsonObject payloadObj = new JsonObject();
            payloadObj.addProperty("channel" , channel);
            payloadObj.addProperty("username" , "TeamCity");
            payloadObj.addProperty("text", message);
            payloadObj.addProperty("icon_url",iconUrl);

            JsonArray attachmentsObj = new JsonArray();
            JsonObject attachment = new JsonObject() ;
            attachmentsObj.add(attachment);

            JsonArray fields = new JsonArray();


            //Attach the build number

            JsonObject buildField = new JsonObject() ;
            buildField.addProperty("title" , "Build");
            buildField.addProperty("short" , true);
            buildField.addProperty("value" , build.getBuildNumber());
            fields.add(buildField);

            StringBuilder fallbackMessage = new StringBuilder();

            if( committersMsg.length() > 0 )
            {
                JsonObject field = new JsonObject() ;

                field.addProperty("title","Changes By");
                field.addProperty("value",committersMsg);
                field.addProperty("short", true);

                fields.add(field);

                fallbackMessage.append("Changes by ");
                fallbackMessage.append(committersMsg);
                fallbackMessage.append(" ");

            }


            //Do we have any issues?

            if( build.isHasRelatedIssues() )
            {

                //We do!
                Collection<Issue> issues = build.getRelatedIssues();

                StringBuilder issueIds = new StringBuilder();
                StringBuilder clickableIssueIds = new StringBuilder();

                for( Issue issue : issues )
                {
                    issueIds.append(',');
                    issueIds.append(issue.getId());

                    clickableIssueIds.append(',');

                    clickableIssueIds.append('<');
                    clickableIssueIds.append(issue.getUrl());
                    clickableIssueIds.append('|');
                    clickableIssueIds.append(issue.getId());
                    clickableIssueIds.append('>');
                }

                if( issueIds.length() > 0 )
                {
                    issueIds.deleteCharAt(0); //delete first ','
                }

                if( clickableIssueIds.length() > 0 )
                {
                    clickableIssueIds.deleteCharAt(0); //delete first ','
                }

               JsonObject field = new JsonObject() ;

                field.addProperty("title", "Related Issues");
                field.addProperty("value",clickableIssueIds.toString());
                field.addProperty("short", true);

                fields.add(field);

                fallbackMessage.append("Related Issues ");
                fallbackMessage.append(issueIds.toString());
            }

            attachment.addProperty("color", (goodColor ? "good" : "danger"));
            attachment.add("fields",fields);

            JsonObject commitAttachment = createCommitAttachment(build);

            if(commitAttachment != null) {
                attachmentsObj.add(commitAttachment);
            }

            if( attachmentsObj.size() > 0 ) {
                payloadObj.add("attachments", attachmentsObj);
            }

            HttpURLConnection conn = (HttpURLConnection) url.openConnection();
            conn.setDoOutput(true);

            BufferedOutputStream bos = new BufferedOutputStream(conn.getOutputStream());

            String payloadJson = getGson().toJson(payloadObj);
            String bodyContents = "payload=" + payloadJson ;
            bos.write(bodyContents.getBytes("utf8"));
            bos.flush();
            bos.close();

            int serverResponseCode = conn.getResponseCode() ;

            conn.disconnect();
            conn = null ;
            url = null ;

        }
        catch ( MalformedURLException ex )
        {

        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private JsonObject createCommitAttachment(SRunningBuild build) {

        List<SVcsModification>  changes =  build.getChanges(SelectPrevBuildPolicy.SINCE_LAST_SUCCESSFULLY_FINISHED_BUILD, true);

        StringBuilder commitMessage = new StringBuilder();

        /*
         * If this field starts to feel too long in slack, we should only use the first item in the array, which would be the latest change
         *
         */
        for ( int i = 0 ; i < changes.size() ; i++ ){
            SVcsModification modification = changes.get(i);
            String desc = modification.getDescription();
            commitMessage.append("‣ ");
            commitMessage.append(desc);

            if( i < changes.size() - 1 ) {
                commitMessage.append("\n");
            }
        }

        if (changes.size() < 1) {
            return null;
        }


        String commitMessageString = commitMessage.toString();
        JsonObject attachment = new JsonObject();
        attachment.addProperty("title", "Commit Messages");
        attachment.addProperty("fallback" , commitMessageString);
        attachment.addProperty("text" , commitMessageString);
        attachment.addProperty("color" , "#2FA8B9");

        Branch branch = build.getBranch();
        if (branch != null) {
            attachment.addProperty("footer" , String.format("Built from %s", branch.getDisplayName()) );

            Date finishDate = build.getFinishDate();
            if (finishDate != null ) {
                long finishTime = finishDate.getTime()/1000;
                attachment.addProperty("ts" , finishTime);
            }
        }

        return attachment;

    }

}

File: src/main/kotlin/com/tapadoo/slacknotifier/SlackServerAdapter.kt
package com.tapadoo.slacknotifier

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.JsonArray
import com.google.gson.JsonObject
import com.tapadoo.slacknotifier.data.SlackField
import jetbrains.buildServer.serverSide.*
import jetbrains.buildServer.serverSide.settings.ProjectSettingsManager
import jetbrains.buildServer.vcs.SelectPrevBuildPolicy
import org.joda.time.Duration
import org.joda.time.format.PeriodFormatterBuilder

import java.io.BufferedOutputStream
import java.io.IOException
import java.net.HttpURLConnection
import java.net.MalformedURLException
import java.net.URL

/**
 * @author Jason Connery
 * @since 04/03/2014
 */
class SlackServerAdapter(
    private val buildServer: SBuildServer,
    private val projectManager: ProjectManager,
    private val projectSettingsManager: ProjectSettingsManager,
    private val slackConfig: SlackConfigProcessor) : BuildServerAdapter() {

  private val gson: Gson by lazy { GsonBuilder().create() }

  fun init() {
    buildServer.addListener(this)
  }

  override fun buildStarted(build: SRunningBuild) {
    super.buildStarted(build)

    if (slackConfig.postStarted) postStartedBuild(build)
  }

  override fun buildFinished(build: SRunningBuild) {
    super.buildFinished(build)

    when {
      build.buildStatus.isSuccessful -> processSuccessfulBuild(build)
      build.buildStatus.isFailed -> postFailureBuild(build)
      else -> {
        //TODO - modify in future if we care about other states
      }
    }
  }

  private fun postStartedBuild(build: SRunningBuild) {
    // Could put other into here. Agents maybe?
    val message = "Project '${build.fullName}' build started."
    postToSlack(build, message, true)
  }

  private fun postFailureBuild(build: SRunningBuild) {
    if (!slackConfig.postFailed) return

    val durationFormatter = PeriodFormatterBuilder()
        .printZeroRarelyFirst()
        .appendHours()
        .appendSuffix(" hour", " hours")
        .appendSeparator(" ")
        .printZeroRarelyLast()
        .appendMinutes()
        .appendSuffix(" minute", " minutes")
        .appendSeparator(" and ")
        .appendSeconds()
        .appendSuffix(" second", " seconds")
        .toFormatter()

    val buildDuration = Duration(1000 * build.duration)
    val message = "Project '${build.fullName}' build failed! (${durationFormatter.print(buildDuration.toPeriod())})"
    postToSlack(build, message, false)
  }

  private fun processSuccessfulBuild(build: SRunningBuild) {
    if (!slackConfig.postSuccessful) return

    val durationFormatter = PeriodFormatterBuilder()
        .printZeroRarelyFirst()
        .appendHours()
        .appendSuffix(" hour", " hours")
        .appendSeparator(" ")
        .printZeroRarelyLast()
        .appendMinutes()
        .appendSuffix(" minute", " minutes")
        .appendSeparator(" and ")
        .appendSeconds()
        .appendSuffix(" second", " seconds")
        .toFormatter()

    val buildDuration = Duration(1000 * build.duration)
    val message = "Project '${build.fullName}' built successfully in ${durationFormatter.print(buildDuration.toPeriod())}."

    postToSlack(build, message, true)
  }

  /**
   * Post a payload to slack with a message and good/bad color. Committer summary is automatically added as an attachment
   *
   * @param build the build the message is relating to
   * @param message main message to include, 'Build X completed...' etc
   * @param goodColor true for 'good' builds, false for danger.
   */
  private fun postToSlack(build: SRunningBuild, message: String, goodColor: Boolean) {
    val postToSlack = build.parametersProvider.get("system.POST_TO_SLACK")
    if (build.isPersonal || postToSlack.equals("false", true) || postToSlack == "0") return

    try {
      val url = URL(slackConfig.postUrl)
      val projectSettings = projectSettingsManager.getSettings(build.projectId, "slackSettings") as SlackProjectSettings

      if (!projectSettings.isEnabled) return

      val iconUrl = projectSettings.logoUrl.takeIf { !it.isNullOrEmpty() } ?: slackConfig.logoUrl
      val configuredChannel = build.parametersProvider.get("system.SLACK_CHANNEL")
      val channel = if (!configuredChannel.isNullOrEmpty()) configuredChannel
      else if (!projectSettings.channel.isNullOrEmpty()) projectSettings.channel
      else slackConfig.defaultChannel

      val committersString = StringBuilder()
      build.getCommitters(SelectPrevBuildPolicy.SINCE_LAST_BUILD).users.forEach { committer ->
        val committerName = committer?.name?.takeIf { it.isNotEmpty() } ?: committer.username
        if (!committerName.isNullOrEmpty()) committersString.append("$committerName, ")
      }

      if (committersString.isNotEmpty()) committersString.deleteCharAt(committersString.length - 1) // remove the last ,
      val committersMsg = committersString.toString()

      val payloadObj = JsonObject().apply {
        addProperty("channel", channel)
        addProperty("username", "TeamCity")
        addProperty("text", message)
        addProperty("icon_url", iconUrl)
      }

      val attachmentsObj = JsonArray()
      val attachment = JsonObject()
      attachmentsObj.add(attachment)

      val fields = JsonArray()

      // Attach the build number
      fields.add(SlackField("Build", build.buildNumber, true).toJson())

      val fallbackMessage = StringBuilder()

      if (committersMsg.isNotEmpty()) {
        fields.add(SlackField("Changes By", committersMsg, true).toJson())
        fallbackMessage.append("Changes by $committersMsg ")
      }

      val testStats = build.getBuildStatistics(BuildStatisticsOptions.ALL_TESTS_NO_DETAILS)
      if (testStats.allTestRunCount > 0) {
        val failureCount = testStats.failedTestCount
        val successCount = testStats.passedTestCount

        fields.add(SlackField("Tests", "$successCount passed, $failureCount failed.", true).toJson())
      }

      // Do we have any issues?
      if (build.isHasRelatedIssues) {

        // We do!
        //val issues = build.relatedIssues.distinctBy { it.id } 
        val issues = build.relatedIssues
        val issueIds = StringBuilder()
        val clickableIssueIds = StringBuilder()

        issues.forEach { issue ->
          issueIds.append("${issue.id}, ")
          clickableIssueIds.append("<${issue.url}|${issue.id}>, ")
        }

        issueIds.dropLastWhile { it == ',' }
        clickableIssueIds.dropLastWhile { it == ',' }

        fields.add(SlackField("Related Issues", clickableIssueIds.toString(), true).toJson())
        fallbackMessage.append("Related Issues $issueIds")
      }

      attachment.addProperty("color", if (goodColor) "good" else "danger")
      attachment.add("fields", fields)

      val commitAttachment = createCommitAttachment(build)
      if (commitAttachment != null) attachmentsObj.add(commitAttachment)
      if (attachmentsObj.size() > 0) payloadObj.add("attachments", attachmentsObj)

      val conn = url.openConnection() as HttpURLConnection
      conn.doOutput = true

      BufferedOutputStream(conn.outputStream).use {
        val bodyContents = "payload=${gson.toJson(payloadObj)}"
        it.write(bodyContents.toByteArray(charset("utf8")))
      }

      conn.responseCode
      conn.disconnect()
    } catch (e: MalformedURLException) {
      e.printStackTrace()
    } catch (e: IOException) {
      e.printStackTrace()
    }
  }

  private fun createCommitAttachment(build: SRunningBuild): JsonObject? {
    val ignoreCommitMessage = build.parametersProvider.get("system.SLACK_IGNORE_COMMIT_MESSAGE")
    if (ignoreCommitMessage.equals("true", ignoreCase = true) || ignoreCommitMessage == "1") return null

    val changes = build.getChanges(SelectPrevBuildPolicy.SINCE_LAST_SUCCESSFULLY_FINISHED_BUILD, true)
    val commitMessage = StringBuilder()

    /*
     * If this field starts to feel too long in slack, we should only use the first item in the array, which would be the latest change
     */
    changes.indices.forEach { i ->
      commitMessage.append("‣ ${changes[i].description}")
      if (i < changes.size - 1) commitMessage.append("\n")
    }

    if (changes.size < 1) return null

    val commitMessageString = commitMessage.toString()
    val attachment = JsonObject().apply {
      addProperty("title", "Commit Messages")
      addProperty("fallback", commitMessageString)
      addProperty("text", commitMessageString)
      addProperty("color", "#2FA8B9")
    }

    build.branch?.let { branch ->
      attachment.addProperty("footer", "Built from ${branch.displayName}")

      build.finishDate?.let { finishDate ->
        attachment.addProperty("ts", finishDate.time / 1000)
      }
    }

    return attachment
  }
}
